src/storage/mod.rs

```diff
@@ -129,6 +140,145 @@ impl Storage {
         Ok(blob)
     }
 
+    pub(super) fn get_range(
+        &self,
+        path: &str,
+        max_size: usize,
+        range: FileRange,
+        compression: Option<CompressionAlgorithm>,
+    ) -> Result<Blob, Error> {
+        let mut blob = match &self.backend {
+            StorageBackend::Database(db) => db.get(path, max_size, Some(range)),
+            StorageBackend::S3(s3) => s3.get(path, max_size, Some(range)),
+        }?;
+        // file content encoding is ignored for ranges
+        // since we only have a range anyways, and we need the encoding
+        // for the range, not the file
+        if let Some(alg) = compression {
+            blob.content = decompress(blob.content.as_slice(), alg, max_size)?;
+            blob.compression = None;
+        }
+        Ok(blob)
+    }
+
+    fn get_index_for(&self, archive_path: &str) -> Result<archive_index::Index, Error> {
+        let remote_index_path = format!("{}.index", archive_path);
+        let local_index_path = self.local_archive_cache_path.join(&remote_index_path);
+
+        if local_index_path.exists() {
+            let mut file = fs::File::open(local_index_path)?;
+            archive_index::Index::load(&mut file)
+        } else {
+            let index_content = self.get(&remote_index_path, std::usize::MAX)?.content;
+
+            fs::create_dir_all(
+                local_index_path
+                    .parent()
+                    .ok_or_else(|| err_msg("index path without parent"))?,
+            )?;
+            let mut file = fs::File::create(&local_index_path)?;
+            file.write_all(&index_content)?;
+
+            archive_index::Index::load(&mut &index_content[..])
```

Comment:
that gives me: 
```
error[E0277]: the trait bound `[u8]: std::io::Read` is not satisfied
   --> src/storage/mod.rs:183:40
    |
183 |             archive_index::Index::load(&mut index_content[..])
    |                                        ^^^^^^^^^^^^^^^^^^^^^^ the trait `std::io::Read` is not implemented for `[u8]`
    |
   ::: src/storage/archive_index.rs:30:30
    |
30  |     pub fn load(reader: impl io::Read) -> Result<Index> {
    |                              -------- required by this bound in `archive_index::Index::load`
    |
    = help: the following implementations were found:
              <&[u8] as std::io::Read>
    = note: required because of the requirements on the impl of `std::io::Read` for `&mut [u8]`

```

Reply:
Huh, that's a weird interaction. Ok, this code seems fine.